src/theory/quantifiers/sygus/example_min_eval.h

```diff
@@ -0,0 +1,124 @@
+/*********************                                                        */
+/*! \file example_min_eval.h
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief utility for minimizing the number of calls to evaluate a term
+ ** on substitutions with a fixed domain.
+ **/
+
+#include "cvc4_private.h"
+
+#ifndef CVC4__THEORY__QUANTIFIERS__EXAMPLE_MIN_EVAL_H
+#define CVC4__THEORY__QUANTIFIERS__EXAMPLE_MIN_EVAL_H
+
+#include <vector>
+
+#include "expr/node.h"
+#include "expr/node_trie.h"
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+/**
+ * Virtual evaluator class for Example Minimize Eval.
+ */
+class EmeEval
+{
+ public:
+  EmeEval() {}
+  virtual ~EmeEval() {}
+  /** Evaluate n given substitution { args -> vals }. */
+  virtual Node eval(TNode n,
+                    const std::vector<Node>& args,
+                    const std::vector<Node>& vals) = 0;
+};
+
+/**
+ * Example Minimize Eval
+ *
+ * This class is designed to evaluate a term on a set of substitutions
+ * with a fixed domain.
+ *
+ * Its key feature is that substitutions that are identical over the free
+ * variables of the term are not recomputed. For example, say I wish to evaluate
+ * x+5 on substitutions having the domain { x, y }. Then, for the substitutions:
+ *  { x -> 0, y -> 1 }
+ *  { x -> 0, y -> 2 }
+ *  { x -> 0, y -> 3 }
+ *  { x -> 1, y -> 3 }
+ * I would only compute the result of 0+5 once. On the other hand, evaluating
+ * x+y for the above substitutions would require 4 evaluations.
+ *
+ * To use this class, call initialize(n,vars) first and then
+ * evaluate(subs_1) ... evaluate(subs_n) as desired. Details on these methods
+ * can be found below.
+ */
+class ExampleMinEval
+{
+ public:
+  ExampleMinEval() {}
+  ~ExampleMinEval() {}
+  /**
+   * Initialize this cache to evaluate n on substitutions with domain vars.
+   * Argument ece is the evaluator object.
+   */
+  void initialize(Node n, const std::vector<Node>& vars, EmeEval* ece);
```

Comment:
Not a huge fan of this empty constructor & initialize pattern as you know :) Is this really needed?

Reply:
This one isn't, I've updated the code.